diff --git a/CMakeLists.txt b/CMakeLists.txt
index 50ea31a..518c207 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -300,12 +300,18 @@ else()
             add_subdirectory(src/tbbmalloc_proxy)
         endif()
     endif()
-    if (NOT BUILD_SHARED_LIBS)
+    if (NOT BUILD_SHARED_LIBS OR SKBUILD)
         message(STATUS "TBBBind build targets are disabled due to unsupported environment")
     else()
         add_subdirectory(src/tbbbind)
     endif()
     if (TBB_INSTALL)
+        if(SKBUILD)
+            set(INSTALL_CONFIG "${CMAKE_INSTALL_DATAROOTDIR}")
+        else()
+            set(INSTALL_CONFIG "${CMAKE_INSTALL_LIBDIR}")
+        endif()
+
         # -------------------------------------------------------------------
         # Installation instructions
         include(CMakePackageConfigHelpers)
@@ -316,7 +322,7 @@ else()
 
         install(EXPORT ${PROJECT_NAME}Targets
                 NAMESPACE TBB::
-                DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
+                DESTINATION ${INSTALL_CONFIG}/cmake/${PROJECT_NAME}
                 COMPONENT devel)
         file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake
                    "include(\${CMAKE_CURRENT_LIST_DIR}/${PROJECT_NAME}Targets.cmake)\n")
@@ -330,7 +336,7 @@ else()
 
         install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake"
                       "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
-                DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
+                DESTINATION ${INSTALL_CONFIG}/cmake/${PROJECT_NAME}
                 COMPONENT devel)
 
         install(FILES "README.md"
